File: Interviews/GFG/Daily-Problems/2022/October/Day-26/The-Smurfs.py
class Solution:
    def getChar (self, a, b):
        if a == 'R':
            return 'B' if b == 'G' else 'G'
        elif a == 'G':
            return 'R' if b == 'B' else 'B'
        return 'R' if b == 'G' else 'G'
        
    def minFind (self, n, a):
        st, ans = [], 0
        for i in a:
            cur = i
            while st and st[-1] != cur:
                cur = self.getChar(st[-1], cur)
                st.pop()
            st.append(cur)
            
        ans = len(st)
        st.clear()
        for i in reversed(a):
            cur = i
            while st and st[-1] != cur:
                cur = self.getChar(st[-1], cur)
                st.pop()
            st.append(cur)
        
        ans = min(ans, len(st))
        return ans if ans % 2 == 0 else 1

File: Interviews/GFG/random-problems/Medium/The-Smurfs/The-Smurfs.cpp
class Solution {
public:
    char getChar (char a, char b) {
        if (a == 'R') {
            return b == 'G' ? 'B' : 'G';
        } else if (a == 'G') {
            return b == 'R' ? 'B' : 'R';
        }
        return b == 'R' ? 'G' : 'R';
    }
    int findMin(int n, char a[]) {
        vector<char> st;
        for (int i = 0; i < n; i++) {
            char cur = a[i];
            while (st.size() and st.back() != cur) {
                cur = getChar (st.back(), cur);
                st.pop_back();
            }
            st.push_back(cur);
        }

        int ans = st.size();
        st.clear();

        for (int i = n - 1; i >= 0; i--) {
            char cur = a[i];
            while (st.size() and st.back() != cur) {
                cur = getChar (st.back(), cur);
                st.pop_back();
            }
            st.push_back(cur);
        }

        ans = min(ans, (int)st.size());
        return ans % 2 == 1 ? 1 : ans;
    }
};